### DISCUSSION TITLE
Conditional Middleware

### DISCUSSION THREAD

**Discussion by MasteryOfMastering** [salvo-rs/salvo#216] [Ideas]
Is it possible to have conditional middleware prior to executing the main handler, something like request.next?

**Comment by MasteryOfMastering** [comment]
This seems to work.

Maybe add a section on the website under middleware as custom middleware?

```#[handler]
async fn auth_middleware(depot: &mut Depot, res: &mut Response, ctrl: &mut FlowCtrl) {
    if let Some(session) = depot.session() {
        if let Some(id) = session.get::<i64>("id") {
            // do some processing here
           return;
        }
    }

    res.set_status_code(StatusCode::UNAUTHORIZED);
    ctrl.skip_rest();
}

**Comment by chrislearn** [comment]
This seems to be a relatively common functional requirement， so I added `hoop_when`, it will wrap the middleware, the middleware will executed when filter returns true.

https://github.com/salvo-rs/salvo/blob/main/crates/core/src/routing/router.rs#L223-L231

**Reply by MasteryOfMastering** [reply]
This is excellent.

Nice work on the framework, slowly become my favourite one across all languages!

**Comment by josejachuf** [comment]
Hi @chrislearn 

How can I implement Basic Auth or another middleware in my handler so that it is only applied when necessary, and the need for this authentication is determined dynamically within the request, for example, when retrieving information from the database?
